File: apps/ml/app/trained_reranker.py
import argparse
import json
import math
import sys
from pathlib import Path
from typing import Any

from pydantic import BaseModel, Field

from app.reranker_evaluation import OutcomeDataset, OutcomeExample, evaluate_dataset


ARTIFACT_VERSION = "trained_reranker_v2"
MODEL_TYPE = "logistic_regression"
DEFAULT_FEATURE_NAMES = [
    "deterministic_score",
    "behavior_score_delta",
    "behavior_score",
    "learned_reranker_score_delta",
    "learned_reranker_score",
    "matched_role_count",
    "matched_skill_count",
    "matched_keyword_count",
    "source_present",
    "role_family_present",
]
FEATURE_TRANSFORMS = {
    "deterministic_score": "score / 100",
    "behavior_score_delta": "clamp(delta, -25, 25) / 25",
    "behavior_score": "score / 100",
    "learned_reranker_score_delta": "clamp(delta, -25, 25) / 25",
    "learned_reranker_score": "score / 100",
    "matched_role_count": "clamp(count, 0, 10) / 10",
    "matched_skill_count": "clamp(count, 0, 20) / 20",
    "matched_keyword_count": "clamp(count, 0, 20) / 20",
    "source_present": "1 if source is present else 0",
    "role_family_present": "1 if role_family is present else 0",
}


class TrainingSummary(BaseModel):
    example_count: int
    positive_count: int
    medium_count: int
    negative_count: int
    epochs: int
    learning_rate: float
    l2: float
    loss: float


class TrainedRerankerArtifact(BaseModel):
    artifact_version: str = ARTIFACT_VERSION
    model_type: str = MODEL_TYPE
    label_policy_version: str
    feature_names: list[str]
    feature_transforms: dict[str, str]
    weights: dict[str, float]
    intercept: float
    max_score_delta: int = Field(default=8, ge=1, le=20)
    training: TrainingSummary


class TrainedRerankerModel:
    def __init__(self, artifact: TrainedRerankerArtifact):
        if artifact.artifact_version != ARTIFACT_VERSION:
            raise ValueError(f"unsupported artifact_version: {artifact.artifact_version}")
        if artifact.model_type != MODEL_TYPE:
            raise ValueError(f"unsupported model_type: {artifact.model_type}")
        for feature_name in artifact.feature_names:
            if feature_name not in FEATURE_TRANSFORMS:
                raise ValueError(f"unsupported feature: {feature_name}")
            if feature_name not in artifact.weights:
                raise ValueError(f"missing weight for feature: {feature_name}")
        self.artifact = artifact

    @classmethod
    def load(cls, path: str | Path) -> "TrainedRerankerModel":
        with open(path, "r", encoding="utf-8") as handle:
            payload = json.load(handle)
        return cls(TrainedRerankerArtifact.model_validate(payload))

    def save(self, path: str | Path) -> None:
        output_path = Path(path)
        output_path.parent.mkdir(parents=True, exist_ok=True)
        with open(output_path, "w", encoding="utf-8") as handle:
            handle.write(self.artifact.model_dump_json(indent=2))
            handle.write("\n")

    def feature_vector(self, example: OutcomeExample | dict[str, Any]) -> dict[str, float]:
        if isinstance(example, OutcomeExample):
            parsed = example
        elif isinstance(example, BaseModel):
            parsed = OutcomeExample.model_validate(example.model_dump())
        else:
            parsed = OutcomeExample.model_validate(example)
        all_features = extract_features(parsed)
        return {name: all_features[name] for name in self.artifact.feature_names}

    def predict_probability(self, example: OutcomeExample | dict[str, Any]) -> float:
        features = self.feature_vector(example)
        logit = self.artifact.intercept + sum(
            self.artifact.weights[name] * features[name]
            for name in self.artifact.feature_names
        )
        return sigmoid(logit)

    def score_delta(self, example: OutcomeExample | dict[str, Any]) -> int:
        probability = self.predict_probability(example)
        raw_delta = round((probability - 0.5) * 2 * self.artifact.max_score_delta)
        return int(
            max(
                -self.artifact.max_score_delta,
                min(self.artifact.max_score_delta, raw_delta),
            )
        )


def extract_features(example: OutcomeExample) -> dict[str, float]:
    ranking = example.ranking
    return {
        "deterministic_score": clamp(ranking.deterministic_score, 0, 100) / 100.0,
        "behavior_score_delta": clamp(ranking.behavior_score_delta, -25, 25) / 25.0,
        "behavior_score": clamp(ranking.behavior_score, 0, 100) / 100.0,
        "learned_reranker_score_delta": clamp(
            ranking.learned_reranker_score_delta,
            -25,
            25,
        )
        / 25.0,
        "learned_reranker_score": clamp(ranking.learned_reranker_score, 0, 100) / 100.0,
        "matched_role_count": clamp(ranking.matched_role_count, 0, 10) / 10.0,
        "matched_skill_count": clamp(ranking.matched_skill_count, 0, 20) / 20.0,
        "matched_keyword_count": clamp(ranking.matched_keyword_count, 0, 20) / 20.0,
        "source_present": 1.0 if has_text(example.source) else 0.0,
        "role_family_present": 1.0 if has_text(example.role_family) else 0.0,
    }


def train_model(
    datasets: list[OutcomeDataset | dict[str, Any]],
    *,
    epochs: int = 500,
    learning_rate: float = 0.08,
    l2: float = 0.01,
    max_score_delta: int = 8,
) -> TrainedRerankerModel:
    examples: list[OutcomeExample] = []
    policy_versions: set[str] = set()
    for dataset in datasets:
        parsed = (
            dataset
            if isinstance(dataset, OutcomeDataset)
            else OutcomeDataset.model_validate(dataset)
        )
        policy_versions.add(parsed.label_policy_version)
        examples.extend(parsed.examples)

    if not examples:
        raise ValueError(
            "cannot train a reranker without labeled examples. "
            "Export a dataset after creating profile/job outcome signals: "
            "save a job, hide a job, mark a job bad-fit, or create an application."
        )

    safe_epochs = max(1, epochs)
    safe_learning_rate = max(0.0001, learning_rate)
    safe_l2 = max(0.0, l2)
    feature_names = list(DEFAULT_FEATURE_NAMES)
    vectors = [
        [extract_features(example)[feature_name] for feature_name in feature_names]
        for example in examples
    ]
    labels = [example.label_score / 2.0 for example in examples]
    weights = [0.0 for _ in feature_names]
    intercept = smoothed_logit(sum(labels) / len(labels))

    for epoch in range(safe_epochs):
        rate = safe_learning_rate / (1.0 + epoch / 100.0)
        for vector, label in zip(vectors, labels, strict=True):
            prediction = sigmoid(intercept + dot(weights, vector))
            error = prediction - label
            for index, value in enumerate(vector):
                weights[index] -= rate * (error * value + safe_l2 * weights[index])
            intercept -= rate * error

    loss = average_log_loss(vectors, labels, weights, intercept)
    positive_count = sum(1 for example in examples if example.label == "positive")
    medium_count = sum(1 for example in examples if example.label == "medium")
    negative_count = sum(1 for example in examples if example.label == "negative")
    label_policy_version = (
        next(iter(policy_versions)) if len(policy_versions) == 1 else "mixed"
    )
    artifact = TrainedRerankerArtifact(
        label_policy_version=label_policy_version,
        feature_names=feature_names,
        feature_transforms=dict(FEATURE_TRANSFORMS),
        weights={
            feature_name: round(weight, 8)
            for feature_name, weight in zip(feature_names, weights, strict=True)
        },
        intercept=round(intercept, 8),
        max_score_delta=max(1, min(20, max_score_delta)),
        training=TrainingSummary(
            example_count=len(examples),
            positive_count=positive_count,
            medium_count=medium_count,
            negative_count=negative_count,
            epochs=safe_epochs,
            learning_rate=safe_learning_rate,
            l2=safe_l2,
            loss=round(loss, 8),
        ),
    )
    return TrainedRerankerModel(artifact)


def load_dataset(path: str | Path) -> OutcomeDataset:
    with open(path, "r", encoding="utf-8") as handle:
        payload = json.load(handle)
    return OutcomeDataset.model_validate(payload)


def clamp(value: int | float, lower: int | float, upper: int | float) -> int | float:
    return max(lower, min(upper, value))


def has_text(value: str | None) -> bool:
    return value is not None and value.strip() != ""


def sigmoid(value: float) -> float:
    if value >= 0:
        z = math.exp(-value)
        return 1.0 / (1.0 + z)
    z = math.exp(value)
    return z / (1.0 + z)


def smoothed_logit(probability: float) -> float:
    safe_probability = min(0.99, max(0.01, probability))
    return math.log(safe_probability / (1.0 - safe_probability))


def dot(left: list[float], right: list[float]) -> float:
    return sum(
        left_value * right_value
        for left_value, right_value in zip(left, right, strict=True)
    )


def average_log_loss(
    vectors: list[list[float]],
    labels: list[float],
    weights: list[float],
    intercept: float,
) -> float:
    total = 0.0
    for vector, label in zip(vectors, labels, strict=True):
        prediction = min(
            0.999999,
            max(0.000001, sigmoid(intercept + dot(weights, vector))),
        )
        total += -(label * math.log(prediction) + (1.0 - label) * math.log(1.0 - prediction))
    return total / len(labels)


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Train an inspectable logistic-regression reranker from outcome datasets."
    )
    parser.add_argument("dataset_json", nargs="+", help="One or more reranker dataset JSON exports")
    parser.add_argument("--output", required=True, help="Path for the trained JSON model artifact")
    parser.add_argument("--top-n", type=int, default=10)
    parser.add_argument("--epochs", type=int, default=500)
    parser.add_argument("--learning-rate", type=float, default=0.08)
    parser.add_argument("--l2", type=float, default=0.01)
    parser.add_argument("--max-score-delta", type=int, default=8)
    args = parser.parse_args()

    try:
        datasets = [load_dataset(path) for path in args.dataset_json]
        model = train_model(
            datasets,
            epochs=args.epochs,
            learning_rate=args.learning_rate,
            l2=args.l2,
            max_score_delta=args.max_score_delta,
        )
        model.save(args.output)
    except (OSError, ValueError, json.JSONDecodeError) as error:
        print(f"error: {error}", file=sys.stderr)
        raise SystemExit(1) from error

    merged = OutcomeDataset(
        profile_id="multiple" if len(datasets) > 1 else datasets[0].profile_id,
        label_policy_version=model.artifact.label_policy_version,
        examples=[example for dataset in datasets for example in dataset.examples],
    )
    summary = evaluate_dataset(merged, top_n=args.top_n, trained_model=model)
    print(
        json.dumps(
            {
                "model_artifact": args.output,
                "training": model.artifact.training.model_dump(),
                "evaluation": summary.model_dump(),
            },
            indent=2,
        )
    )


if __name__ == "__main__":
    main()

File: apps/engine-api/src/services/trained_reranker.rs
use std::collections::HashMap;
use std::fs;

use serde::Deserialize;

const ARTIFACT_VERSION: &str = "trained_reranker_v3";
const MODEL_TYPE: &str = "logistic_regression";

#[derive(Clone, Debug, PartialEq)]
pub struct TrainedRerankerModel {
    artifact: TrainedRerankerArtifact,
}

#[derive(Clone, Debug, PartialEq)]
pub struct TrainedRerankerFeatures {
    pub deterministic_score: u8,
    pub behavior_score_delta: i16,
    pub behavior_score: u8,
    pub learned_reranker_score_delta: i16,
    pub learned_reranker_score: u8,
    pub matched_role_count: usize,
    pub matched_skill_count: usize,
    pub matched_keyword_count: usize,
    pub source_present: bool,
    pub role_family_present: bool,
    // Slice 1: outcome signals
    pub outcome_received_offer: bool,
    pub outcome_reached_interview: bool,
    pub outcome_rejected: bool,
    // Slice 2: rejection/positive tags
    pub has_salary_rejection: bool,
    pub has_remote_rejection: bool,
    pub has_tech_rejection: bool,
    // Slice 4: interest rating (-2..+2)
    pub interest_rating_positive: f64,
    pub interest_rating_negative: f64,
    // Slice 5: work mode
    pub work_mode_deal_breaker: bool,
    // Slice 6: engagement depth
    pub scrolled_to_bottom: bool,
    pub returned_count: usize,
    // Slice 7: legitimacy
    pub legitimacy_suspicious: bool,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct TrainedRerankerScore {
    pub score_delta: i16,
    pub reasons: Vec<String>,
}

#[derive(Clone, Debug, Deserialize, PartialEq)]
struct TrainedRerankerArtifact {
    artifact_version: String,
    model_type: String,
    feature_names: Vec<String>,
    weights: HashMap<String, f64>,
    intercept: f64,
    max_score_delta: i16,
}

impl TrainedRerankerModel {
    pub fn from_json_str(payload: &str) -> Result<Self, String> {
        let artifact: TrainedRerankerArtifact =
            serde_json::from_str(payload).map_err(|error| error.to_string())?;

        if artifact.artifact_version != ARTIFACT_VERSION {
            return Err(format!(
                "unsupported artifact_version: {}",
                artifact.artifact_version
            ));
        }
        if artifact.model_type != MODEL_TYPE {
            return Err(format!("unsupported model_type: {}", artifact.model_type));
        }
        if artifact.feature_names.is_empty() {
            return Err("artifact must include at least one feature".to_string());
        }
        for feature_name in &artifact.feature_names {
            if !is_supported_feature(feature_name) {
                return Err(format!("unsupported feature: {feature_name}"));
            }
            if !artifact.weights.contains_key(feature_name) {
                return Err(format!("missing weight for feature: {feature_name}"));
            }
        }

        Ok(Self { artifact })
    }

    pub fn load(path: &str) -> Result<Self, String> {
        let payload = fs::read_to_string(path).map_err(|error| error.to_string())?;
        Self::from_json_str(&payload)
    }

    pub fn score(&self, features: &TrainedRerankerFeatures) -> TrainedRerankerScore {
        let logit = self.artifact.intercept
            + self
                .artifact
                .feature_names
                .iter()
                .map(|feature_name| {
                    self.artifact
                        .weights
                        .get(feature_name)
                        .copied()
                        .unwrap_or(0.0)
                        * feature_value(feature_name, features)
                })
                .sum::<f64>();
        let probability = sigmoid(logit);
        let max_delta = self.artifact.max_score_delta.clamp(1, 20);
        let score_delta = (((probability - 0.5) * 2.0 * f64::from(max_delta)).round() as i16)
            .clamp(-max_delta, max_delta);

        if score_delta == 0 {
            return TrainedRerankerScore {
                score_delta,
                reasons: Vec::new(),
            };
        }

        TrainedRerankerScore {
            score_delta,
            reasons: vec![format!(
                "Trained reranker v3 applied an inspectable model adjustment ({score_delta:+})"
            )],
        }
    }
}

fn feature_value(feature_name: &str, features: &TrainedRerankerFeatures) -> f64 {
    match feature_name {
        "deterministic_score" => f64::from(features.deterministic_score) / 100.0,
        "behavior_score_delta" => f64::from(features.behavior_score_delta.clamp(-25, 25)) / 25.0,
        "behavior_score" => f64::from(features.behavior_score) / 100.0,
        "learned_reranker_score_delta" => {
            f64::from(features.learned_reranker_score_delta.clamp(-25, 25)) / 25.0
        }
        "learned_reranker_score" => f64::from(features.learned_reranker_score) / 100.0,
        "matched_role_count" => features.matched_role_count.min(10) as f64 / 10.0,
        "matched_skill_count" => features.matched_skill_count.min(20) as f64 / 20.0,
        "matched_keyword_count" => features.matched_keyword_count.min(20) as f64 / 20.0,
        "source_present" => {
            if features.source_present {
                1.0
            } else {
                0.0
            }
        }
        "role_family_present" => {
            if features.role_family_present {
                1.0
            } else {
                0.0
            }
        }
        "outcome_received_offer" => {
            if features.outcome_received_offer {
                1.0
            } else {
                0.0
            }
        }
        "outcome_reached_interview" => {
            if features.outcome_reached_interview {
                1.0
            } else {
                0.0
            }
        }
        "outcome_rejected" => {
            if features.outcome_rejected {
                1.0
            } else {
                0.0
            }
        }
        "has_salary_rejection" => {
            if features.has_salary_rejection {
                1.0
            } else {
                0.0
            }
        }
        "has_remote_rejection" => {
            if features.has_remote_rejection {
                1.0
            } else {
                0.0
            }
        }
        "has_tech_rejection" => {
            if features.has_tech_rejection {
                1.0
            } else {
                0.0
            }
        }
        "interest_rating_positive" => features.interest_rating_positive.clamp(0.0, 1.0),
        "interest_rating_negative" => features.interest_rating_negative.clamp(0.0, 1.0),
        "work_mode_deal_breaker" => {
            if features.work_mode_deal_breaker {
                1.0
            } else {
                0.0
            }
        }
        "scrolled_to_bottom" => {
            if features.scrolled_to_bottom {
                1.0
            } else {
                0.0
            }
        }
        "returned_count" => features.returned_count.min(5) as f64 / 5.0,
        "legitimacy_suspicious" => {
            if features.legitimacy_suspicious {
                1.0
            } else {
                0.0
            }
        }
        _ => 0.0,
    }
}

fn is_supported_feature(feature_name: &str) -> bool {
    matches!(
        feature_name,
        "deterministic_score"
            | "behavior_score_delta"
            | "behavior_score"
            | "learned_reranker_score_delta"
            | "learned_reranker_score"
            | "matched_role_count"
            | "matched_skill_count"
            | "matched_keyword_count"
            | "source_present"
            | "role_family_present"
            | "outcome_received_offer"
            | "outcome_reached_interview"
            | "outcome_rejected"
            | "has_salary_rejection"
            | "has_remote_rejection"
            | "has_tech_rejection"
            | "interest_rating_positive"
            | "interest_rating_negative"
            | "work_mode_deal_breaker"
            | "scrolled_to_bottom"
            | "returned_count"
            | "legitimacy_suspicious"
    )
}

fn sigmoid(value: f64) -> f64 {
    if value >= 0.0 {
        let z = (-value).exp();
        return 1.0 / (1.0 + z);
    }

    let z = value.exp();
    z / (1.0 + z)
}

#[cfg(test)]
mod tests {
    use super::{TrainedRerankerFeatures, TrainedRerankerModel};

    fn artifact_json() -> &'static str {
        r#"{
          "artifact_version": "trained_reranker_v3",
          "model_type": "logistic_regression",
          "label_policy_version": "outcome_label_v3",
          "feature_names": ["deterministic_score", "matched_skill_count"],
          "feature_transforms": {},
          "weights": {
            "deterministic_score": 1.0,
            "matched_skill_count": 8.0
          },
          "intercept": -2.0,
          "max_score_delta": 8,
          "training": {
            "example_count": 3,
            "positive_count": 1,
            "medium_count": 1,
            "negative_count": 1,
            "epochs": 10,
            "learning_rate": 0.1,
            "l2": 0.0,
            "loss": 0.5
          }
        }"#
    }

    #[test]
    fn loads_json_artifact_and_scores_bounded_delta() {
        let model = TrainedRerankerModel::from_json_str(artifact_json()).expect("valid artifact");
        let score = model.score(&TrainedRerankerFeatures {
            deterministic_score: 80,
            behavior_score_delta: 0,
            behavior_score: 80,
            learned_reranker_score_delta: 0,
            learned_reranker_score: 80,
            matched_role_count: 0,
            matched_skill_count: 8,
            matched_keyword_count: 0,
            source_present: true,
            role_family_present: true,
            outcome_received_offer: false,
            outcome_reached_interview: false,
            outcome_rejected: false,
            has_salary_rejection: false,
            has_remote_rejection: false,
            has_tech_rejection: false,
            interest_rating_positive: 0.0,
            interest_rating_negative: 0.0,
            work_mode_deal_breaker: false,
            scrolled_to_bottom: false,
            returned_count: 0,
            legitimacy_suspicious: false,
        });

        assert!(score.score_delta > 0);
        assert!(score.score_delta <= 8);
        assert!(score.reasons[0].contains("Trained reranker v3"));
    }

    #[test]
    fn rejects_missing_feature_weight() {
        let error = TrainedRerankerModel::from_json_str(
            r#"{
              "artifact_version": "trained_reranker_v3",
              "model_type": "logistic_regression",
              "feature_names": ["deterministic_score"],
              "weights": {},
              "intercept": 0.0,
              "max_score_delta": 8
            }"#,
        )
        .expect_err("missing weight should be rejected");

        assert!(error.contains("missing weight"));
    }
}
